### PR TITLE
[SEMI-MODULAR] [CURSED AS FUCK] Client-Side input window & Typing Indicator refactor

### PR DESCRIPTION
**Pull request by Kapu1178**
## About The Pull Request
Ports my own version of the typing indicator from Beestation, as well as KubeRoot's black magic fixes:
- https://github.com/BeeStation/BeeStation-Hornet/pull/5417

Refactors typing indicator and the `say/me` verbs. What the fuck does this actually do? Well:

Invoking the `say` verb no longer has a ping-based delay, and the typing indicator will *NEVER* stick, it will always remove itself when the client's dialog has closed. As a side effect, the `say` and `me` verbs take on different appearances, but it isn't too jarring.
 

## How This Contributes To The Skyrat Roleplay Experience
Functioning game mechanics good.

## Changelog

:cl: KubeRoot, Kapu1178
refactor: Typing Indicator has been recoded.
fix: Typing Indicators will always vanish if the user closes the input window.
refactor: Say & Me verbs are invoked using some client side magic
add: Client-side input windows (Kuberoot is a wizard)
/:cl:

### CODE REVIEW

**Comment by fighterslam** [comment]
god is dead and byond killed him

**Comment by gonenoculer5** [comment]
Holy fuck finally. I was about to do this if no one did because the beestation version is phenomenally based.

**Comment by thestubborn** [comment]
question

does this work with subtle and so on?
because having indicators for them would be extra based

**Comment by Kapu1178** [comment]
> question
> 
> does this work with subtle and so on? because having indicators for them would be extra based

No, as these are hooked directly into the existing hotkeys for the respective verbs. Would it be possible to add it to subtle and subtler? Probably. But thats a headache I don't want right now.

**Comment by thestubborn** [comment]
> > question
> > does this work with subtle and so on? because having indicators for them would be extra based
> 
> No, as these are hooked directly into the existing hotkeys for the respective verbs. Would it be possible to add it to subtle and subtler? Probably. But thats a headache I don't want right now.

completely understandable mb

**Review by Gandalf2k15 — APPROVED** [review]

**Comment by RenardQueenston** [comment]
The 'say' window on the left is for `F3` or `T` buttons, the other is for the `IC`>`Say` button.
Because of client-side input I cant use `Ctrl+Arrowkey` hotkeys to jump between the words or `Ctrl+Shift+Arrowkey` to jump and highlight text.
Could you fix this please? It was a really helpful feature, at least for me. (I don't want to use notepad again for sanitizing my messages before I can send them)
https://imgur.com/a/iBmACAR

**Comment by RenardQueenston** [comment]
The 'say' window on the left is for `F3` or `T` buttons, the other is for the `IC`>`Say` button.
Because of client-side input I cant use `Ctrl+Arrowkey` hotkeys to jump between the words or `Ctrl+Shift+Arrowkey` to jump and highlight text.
Could you fix this please? It was a really helpful feature, at least for me. (I don't want to use notepad again for sanitizing my messages before I can send them)
https://imgur.com/a/iBmACAR

**Comment by Weblure** [comment]
> I cant use `Ctrl+Arrowkey` hotkeys to jump between the words or `Ctrl+Shift+Arrowkey` to jump and highlight text.

Wait, that was a thing? Holy shit, that would've been useful to know

File: modular_skyrat/modules/indicators/code/typing_indicator.dm [hunk 1]
```diff
@@ -1,24 +1,85 @@
...
+
+When you press the hotkey, the .start_typing verb is called with the source ("say" or "me") to show the typing indicator.
+When you send a message from the custom window, the appropriate verb is called, .say or .me
+If you close the window without actually sending the message, the .cancel_typing verb is called with the source.
+
+Both the say/me wrappers and cancel_typing remove the typing indicator.
+*/
+
+/// Show the typing indicator. The source signifies what action the user is typing for.
+/mob/verb/start_typing(source as text) // The source argument is currently unused
```
Comment (Iamgoofball) [inline modular_skyrat/modules/indicators/code/typing_indicator.dm:38]: holy proc call overhead batman

**Comment by Iamgoofball** [comment]
this isn't going to work, there's massive issues with lagginess by going through the skin.dmf stuff, i'm constantly having issues with pressing T and getting an extra T appended to the beginning of my sentence

**Comment by hugmeorelse** [comment]
I haven't noticed any issues personally. This seems like solid work from my personal perspective.

**Comment by Kapu1178** [comment]
> this isn't going to work, there's massive issues with lagginess by going through the skin.dmf stuff, i'm constantly having issues with pressing T and getting an extra T appended to the beginning of my sentence

This exact code has existed on Beestation for months and nobody has reported any such issue. 

On your own PR, at 0:39 seconds on the video, you have the same problem. Your keyboard is dirty.

**Comment by Iamgoofball** [comment]
>On your own PR, at 0:39 seconds on the video, you have the same problem. Your keyboard is dirty.

No, that was me legitimately pressing T twice. I'll go back on live and test it again and see if I can get some video.

**Comment by Iamgoofball** [comment]
Alright, I went to test on live and I can't reproduce it. You're right that it's probably my end.

File: code/_onclick/click.dm [hunk 2]
```diff
@@ -105,9 +105,6 @@
 		CtrlClickOn(A)
 		return
 
-	//SKYRAT EDIT ADDITION BEGIN - TYPING_INDICATOR
-	if(typing_indicator)
-		set_typing_indicator(FALSE)
 	//SKYRAT EDIT ADDITION END
```
Comment (Vexylius) [inline code/_onclick/click.dm:108]: Disgusting comment leftover

Comment (Vexylius) [reply code/_onclick/click.dm:108]: (You remove the start of the Skyrat edit but you leave the end of it)

**Comment by nikothedude** [comment]
In a brief period of testing this out on live, I found there FEELS like there's a weird delay kind of sort of thing? I dont know what the fuck to call it, but I feel like the text box takes a little longer than what it should take to appear.

Im having the EXACT issue goof is describing.

**Comment by Kapu1178** [comment]
> In a brief period of testing this out on live, I found there FEELS like there's a weird delay kind of sort of thing? I dont know what the fuck to call it, but I feel like the text box takes a little longer than what it should take to appear.
> 
> Im having the EXACT issue goof is describing.

It's invoked using client side verbs. This has been tested using simulated lag via virtual machines, there is 0 delay.

**Comment by tf-4** [comment]
> Alright, I went to test on live and I can't reproduce it. You're right that it's probably my end.

I've had the same issue. It may not just be you.
Also the /me window is small, and you can't navigate with `ctrl+arrow keys` which is pretty annoying.

**Comment by nikothedude** [comment]
> > In a brief period of testing this out on live, I found there FEELS like there's a weird delay kind of sort of thing? I dont know what the fuck to call it, but I feel like the text box takes a little longer than what it should take to appear.
> > Im having the EXACT issue goof is describing.
> 
> It's invoked using client side verbs. This has been tested using simulated lag via virtual machines, there is 0 delay.

Either SOMETHING funky is happening, or multiple people are encountering some kind of mandela effect or whatever its called.

**Comment by Kapu1178** [comment]
> > Alright, I went to test on live and I can't reproduce it. You're right that it's probably my end.
> 
> I've had the same issue. It may not just be you. Also the /me window is small, and you can't navigate with `ctrl+arrow keys` which is pretty annoying.

Me window can (probably) be expanded, i'll look into it. The latter is actually a bug/oversight with skin inputs, I'll reach out to Lummox about it.

**Review by stylemistake — APPROVED** [review]
I like this

**Comment by nikothedude** [comment]
Okay, yeah, after playing a bit there is definitely something wrong with this, I frequently say something than immediately re-open another text box, and this usually causes something like a stray t or * to be in there due to what I presume are inputs I queued. Of course I'm not the best playtester so this isnt the best feedback, but something is definitely wrong.

**Comment by nikothedude** [comment]
And after a bit moer playing, I think I figured out what's happening, there's a brief period after sending a message that you can continue to type in the text box with, and anything put in that text box will carry over to any other text boxes you open. AKA if you try to say something, then try to press t to topen say, it'll just keep that "t" in the buffer, and once you open the saybox again, the T will remain.

**Comment by GoldenAlpharex** [comment]
> And after a bit moer playing, I think I figured out what's happening, there's a brief period after sending a message that you can continue to type in the text box with, and anything put in that text box will carry over to any other text boxes you open. AKA if you try to say something, then try to press t to topen say, it'll just keep that "t" in the buffer, and once you open the saybox again, the T will remain.

Good catch.

**Comment by nikothedude** [comment]
I also found a way to defeat the typing indicator, if you open say, then hit enter and t in very quick succession, the typing indicator will remain despite no text box being open.

**Comment by GoldenAlpharex** [comment]
I'll close this for now, feel free to re-open it later if you manage to fix the issues we discussed in DMs, because I don't want the most used interfaces in the game to be suffering from a regression.

**Pull request closed without merging**